### ISSUE TITLE
"Socket timeout exception" due to frequent cancelation of HTTP2 byte ranged requests

### ISSUE THREAD

**Issue by SukritGupta17** [square/okhttp#5449]
### Issue description
Experiencing "Socket timeout exception" when byte ranged requests are cancelled frequently.

Running a test where the client side requests about ~10mbytes of data using okhttp client with HTTP2 protocol enabled and cancelling the request after receiving 1000 bytes,
produces a socket timeout exception after a few hundred requests and the test fails.

### Reproduction steps
By running the following test code the problem can be reproduced easily, Add the following piece of code
in the `CallTest.java` file found under the okhttp module test directory:

```java
@Test
public void socketTimeoutTest() throws Exception {
  enableProtocol(Protocol.HTTP_2);
  MockResponse mockResponse = new MockResponse().setBody(new String(new char[10 * 1024 * 1024]));

  for (int i = 0; i < 5000; i++) {
    server.enqueue(mockResponse);
    System.out.println("~~~~Running for " + i + "~~~~");
    URL url = server.url("/").url();
    Request request = new Request.Builder().url(url)
            .header("range", "bytes=" + 1 + "-" + 1000000).build();
    Call call = client.newCall(request);
    Response response = call.execute();
    ResponseBody body = response.body();
    if (body == null) throw new IllegalStateException("No available byte stream");
    try (InputStream inputStream = body.byteStream()) {
      System.out.println("response code: " + response.code());
      int counter = 0;
      // read until 1000 bytes, then cancel
      while (counter < 1000) {
        counter += inputStream.read(new byte[1000]);
      }
    }
    System.out.println("request cancelled");
  }
}
```

You will also need to increase the JUnit timeout settings by adding the following line at the top of the class
```java 
@Rule public final TestRule timeout = new Timeout(Long.MAX_VALUE, TimeUnit.MILLISECONDS);
``` 

### Stacktrace printed out

STANDARD_OUT:
```sh
okhttp3.CallTest > cloudFrontOkHttpTest FAILED
    java.net.SocketTimeoutException: timeout
        at okhttp3.internal.http2.Http2Stream$StreamTimeout.newTimeoutException(Http2Stream.kt:701)
        at okhttp3.internal.http2.Http2Stream$StreamTimeout.exitAndThrowIfTimedOut(Http2Stream.kt:710)
        at okhttp3.internal.http2.Http2Stream$FramingSource.read(Http2Stream.kt:398)
        at okhttp3.internal.connection.Exchange$ResponseBodySource.read(Exchange.kt:279)
        at okio.RealBufferedSource$inputStream$1.read(RealBufferedSource.kt:438)
        at java.base/java.io.InputStream.read(InputStream.java:205)
        at okhttp3.CallTest.cloudFrontOkHttpTest(CallTest.java:1201)
```

STANDARD_ERROR:
```sh
okhttp3.CallTest > cloudFrontOkHttpTest STANDARD_ERROR
    java.net.SocketTimeoutException: timeout
    Sep. 11, 2019 4:50:34 PM okhttp3.internal.platform.Platform log
    INFO: Http2Connection.Listener failure for localhost
    java.io.IOException: stream closed
        at okhttp3.internal.http2.Http2Connection.writeData(Http2Connection.kt:314)
        at okhttp3.internal.http2.Http2Stream$FramingSink.emitFrame(Http2Stream.kt:592)
        at okhttp3.internal.http2.Http2Stream$FramingSink.write(Http2Stream.kt:547)
        at okio.RealBufferedSink.emitCompleteSegments(RealBufferedSink.kt:181)
        at okio.RealBufferedSink.write(RealBufferedSink.kt:39)
        at okhttp3.mockwebserver.MockWebServer.throttledTransfer(MockWebServer.kt:868)
        at okhttp3.mockwebserver.MockWebServer.access$throttledTransfer(MockWebServer.kt:100)
        at okhttp3.mockwebserver.MockWebServer$Http2SocketHandler.writeResponse(MockWebServer.kt:1070)
        at okhttp3.mockwebserver.MockWebServer$Http2SocketHandler.onStream(MockWebServer.kt:964)
        at okhttp3.internal.http2.Http2Connection$ReaderRunnable$headers$$inlined$synchronized$lambda$1.run(Util.kt:580)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
        Suppressed: okhttp3.internal.http2.StreamResetException: stream was reset: CANCEL
                at okhttp3.internal.http2.Http2Stream.checkOutNotClosed$okhttp(Http2Stream.kt:673)
                at okhttp3.internal.http2.Http2Stream$FramingSink.emitFrame(Http2Stream.kt:579)
                at okhttp3.internal.http2.Http2Stream$FramingSink.close(Http2Stream.kt:638)
                at okio.RealBufferedSink.close(RealBufferedSink.kt:244)
                at kotlin.io.CloseableKt.closeFinally(Closeable.kt:56)
                at okhttp3.mockwebserver.MockWebServer$Http2SocketHandler.writeResponse(MockWebServer.kt:1068)
                ... 5 more

```

### Version of okhttp3 being used

okhttp-4.2.0

**Comment by swankjesse** [comment]
Took about 4.5 minutes, but it completed normally for me.
```
Running with Jdk9Platform
Sep 11, 2019 11:02:41 PM okhttp3.mockwebserver.MockWebServer$start$$inlined$execute$1 run
INFO: MockWebServer[65012] starting to accept connections
Sep 11, 2019 11:02:41 PM okhttp3.mockwebserver.MockWebServer$start$$inlined$execute$1 run
INFO: MockWebServer[65013] starting to accept connections
~~~~Running for 0~~~~
response code: 200
request cancelled
~~~~Running for 1~~~~
response code: 200
...
request cancelled
~~~~Running for 4999~~~~
response code: 200
request cancelled
Sep 11, 2019 11:07:12 PM okhttp3.mockwebserver.MockWebServer acceptConnections
INFO: MockWebServer[65013] done accepting connections: Socket closed
Sep 11, 2019 11:07:12 PM okhttp3.mockwebserver.MockWebServer acceptConnections
INFO: MockWebServer[65012] done accepting connections: Socket closed
```

**Comment by swankjesse** [comment]
@SukritGupta17 I suspect this was fixed with this change that was released in 4.2:
https://github.com/square/okhttp/pull/5434

And I think your build tools might be fighting you. The line numbers in your stack trace are not from 4.2.

**Comment by SukritGupta17** [comment]
Ah, yes I am still using 4.1, it didn't rebuild for some reason.
The [#5434 ](https://github.com/square/okhttp/pull/5434)  PR is exactly the problem I was seeing while debugging this, there was some sort of leak in bytes causing an inaccurate window update being sent to the server, which eventually lead to the shrinking of window size resulting in a stall,
sort of a deadlock where the sever is waiting for a window update and the client is waiting for a response from the server.

Thanks @swankjesse for the quick reply. I've tested again with 4.2 and the test passes.
